### PR TITLE
Introduce `--show-inputs-only` Flag

### PR DESCRIPTION
**Pull request by dkelosky**
Draft PR to address: https://github.com/zowe/zowe-cli/issues/870.  

The draft name of the flag is `--show-resolved-args` and is a hidden flag.  I wanted to use `--dry-run` but that appears to be taken in `zowe config init` 😢 .  

If I use old profiles which contain connection info, set environmental variables like `set ZOWE_OPT_PASS=fakepass`, and supply command line arguments, `--show-resolved-args` describes what input would be used for the given command without actually executing the command:
image

Or:
```
C:\>zowe endevor list elements --user mrendvr --host endevor.com --show-resolved-args --subsystem notreal
commandValues:
  environment:        dev
  system:             GACODE
  subsystem:          notreal
  stage-number:       1
  maxrc:              12
  data:               all
  host:               endevor.com
  port:               50002
  instance:           WEBSMFNE
  protocol:           http
  user:               mrendvr
  password:           fakepass
  show-resolved-args: true
optionalProfiles:
  - endevor
  - endevor-location
  - base
```

This is a draft only PR.  We can change or close this in it's current form, but I'm hopeful for some feedback.  

Signed-off-by: Dan Kelosky <dkelosky@gmail.com>

### CODE REVIEW

**Comment by MikeBauerCA** [comment]
This would be an excellent debugging tool as you can easily see what values are being used. Oftentimes it is tedious to figure this out as you may need to look at multiple configs and env vars.

If a value is stored in the secure credential store, I would want a "securely stored"-like message to be displayed rather than the value itself. If these values must be printed to output, I think it should made very clear to the end user that securely stored properties are about to be displayed - perhaps with an a prompt to continue. I'd like to get some feedback from the community before merging the PR if the secure values are to be printed to output.

I'm also open to this being a visible flag, allowing users to debug on their own. This decision is also influenced by whether secure values are printed.

**Comment by dkelosky** [comment]
thanks for looking, we should be able to accommodate either the prompt or a flag... I'll experiment and get back.

it turns out, in my case, I wanted to see both user and password because I was getting a http 403.  it turned out that both user and password were unexpected when I used this flag, and so I was able to solve my problem

**Comment by MikeBauerCA** [comment]
@dkelosky another option would be to add a `--prompt` flag with a default value of `true`. May be easier for users to append `--prompt false` than set an env var.

**Comment by dkelosky** [comment]
@MikeBauerCA good thought.  a reason I was avoiding `--no-prompt` would be a global option on every command, but i'd really want a `--no-prompt` flag which only applies to a `--show-resolved-args` flag.

i agree `--no-prompt` is easier for for most interactive users - except for me, i'd put `ZOWE_SHOW_RESOLVED_ARGS=true` as a system environmental variable 😄

**Comment by t1m0thyj** [comment]
This looks really useful for debugging 🙂 A few thoughts:
* Would it be useful to show where each command value came from? Like `env var`, `cli arg`, `profile`, etc. Depending on how advanced this logic is, it could show more specific info like "base profile" vs "service profile", or "base profile from global config" vs "zosmf.lpar1 profile from project config".
* IMO it would be nice if we could use `--dry-run` for the argument name. I wonder if we should consider making a v2-v2 breaking change on `zowe config init` and `zowe config auto-init` that would rename `--dry-run` to something else like `--preview-changes`. Or maybe we could add an optional dry run handler to commands, allowing them to run their own custom code after we execute your default `--dry-run` logic in CommandProcessor.

**Comment by dkelosky** [comment]
I'm with you there.  When we talked about this in the original issue, we planned on `--dry-run`.  I hadn't realized another `--dry-run` had come along in the meantime; however, I do think `--dry-run` would be far more intuitive for this purpose than to only apply to `zowe config`.  @MikeBauerCA - what do you think?


I'll keep your first point in mind when I revisit this.  I forget how much detail I was able to derive at the time this runs to determine things like where arguments came from.  Thanks for the suggestion.  I'll try to get back to this soon!

**Comment by MikeBauerCA** [comment]
@dkelosky @t1m0thyj 
What do you think of `show-input-only`, `show-input`, `input`, or something similar? 

In addition to being a v2-v2 breaking change, I worry that a command like `zowe jobs submit ds "hlq.ex(ample)" --dry-run` would give the impression that perhaps it is checking if the job would run successfully without actually running it. There may be other similar examples. Perhaps it isn't significant but we may be able to use something other than `show-resolved-args` or `dry-run`.

**Comment by dkelosky** [comment]
Here's a stab at `--show-inputs-only`.  By default, hides user name and password the same way prompting for certain command parameters or log messages are masked (as a side note, there does not appear to be a single place where a "secure" items list is retained):
image

When you set the env, the real values are displayed without the red text.

Please let me know what you think, and I'll start finalizing this.

**Comment by dkelosky** [comment]
right now `--show-secure-args` is not an option since it only made sense if `--show-inputs-only` is also present.  loosely, it also seems similar to log level adjustments which are only via env vars.  is that fair?

**Comment by t1m0thyj** [comment]
This looks good to me and I like the name `--show-inputs-only` 🙂

To determine which values should be treated as secure, I believe the default list of secure values is defined in `LoggerUtils.CENSORED_OPTIONS`. I'll defer to @zFernand0 on how to get the list of user-defined secure values from team config.

**Comment by dkelosky** [comment]
I copied this list: https://github.com/zowe/imperative/blob/master/packages/rest/src/session/ConnectionPropsForSessCfg.ts#L294

@zFernand0 WHAT DO

**Comment by zFernand0** [comment]
> how to get the list of user-defined secure values from team config

For the user-defined  secure properties, isn't `config.api.secure.secureFields()` enough? 
It will get you the full path, and if you know the profile name, then doing `config.api.secure.securePropsForProfile("myProfile")` should work.

Maybe I'm missing the point 😋 

> WHAT DO

`¯\_(ツ)_/¯`

**Comment by zFernand0** [comment]
@dkelosky How about this?
https://github.com/zowe/imperative/compare/dry-run-show-args...test-dry-run-branch?expand=1

Result
image

**Comment by dkelosky** [comment]
I can use that util - my question was because I saw this: `private static readonly secureSessCfgProps: string[] = ["user", "password", "tokenValue", "passphrase"];` [here](https://github.com/zowe/imperative/blob/master/packages/rest/src/session/ConnectionPropsForSessCfg.ts#L294).

**Comment by dkelosky** [comment]
I had unpushed changes pending some feedback so I can't use the diff as-is but I'll pull the code in.  I think I'll try to fall back to the list from `ConnectionPropsForSessCfg` in case old profiles are in effect.

**Comment by zFernand0** [comment]
> I can use that util - my question was because I saw this: `private static readonly secureSessCfgProps: string[] = ["user", "password", "tokenValue", "passphrase"];` [here](https://github.com/zowe/imperative/blob/master/packages/rest/src/session/ConnectionPropsForSessCfg.ts#L294).

I knew I was missing something 😋 
What about making that a public property of the class (or introduce a getter `/shrug`) ? 😋

**Comment by dkelosky** [comment]
✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️

**Comment by zFernand0** [comment]
> ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️

@dkelosky , Where are the 2 skipped ones ?
🚫 🚫 ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️ ✔️

**Comment by dkelosky** [comment]
I made a mistake - THANKS FOR NOTICING 😄

**Comment by gejohnston** [comment]
Sorry that I came so late to the table. I agree with the value of this flag, and in the direction that your UX conversation has taken things.

I just have one question: The CHANGELOG refers to this as a "hidden flag". Does this imply that we do not want users to see this flag in our online help?

**Comment by dkelosky** [comment]
I marked the flag as hidden to avoid [what I thought might be] cluttering up the help menu, not to hide it from the users.  I have no strong opinion though - if a visible flag is preferred, I can make that change

File: packages/cmd/src/CommandProcessor.ts
Comment (t1m0thyj) [inline packages/cmd/src/CommandProcessor.ts:807]: Another option for a "censor response" would be `(secure value)` which is what the "config list" command uses.

Not sure which value is preferred here, just mentioning this for consideration 🙂 Perhaps we set a precedent of using `****` in log files and `(secure value)` in CLI output?

File: CHANGELOG.md
Comment (t1m0thyj) [inline CHANGELOG.md:8]: Based on discussion at standup, I think we agreed that making the option unhidden would be preferred, as long as there's an easy way to define it at the top level so that it applies to all commands but not command groups.

@gejohnston @awharn @zFernand0 Do you recall the same outcome of our discussion?

Comment (gejohnston) [reply CHANGELOG.md:8]: Yes Timothy, the visibility of the flag as you described it was what I recall that Mike B advocated. In fact he said he would add that comment to this PR, but I suspect that he got sidetracked.

**Comment by dkelosky** [comment]
Thanks all - I'll try to get these changes "soon"

File: packages/rest/src/session/ConnectionPropsForSessCfg.ts
Comment (dkelosky) [reply packages/rest/src/session/ConnectionPropsForSessCfg.ts]: i blame your CI 😎 - fixed

File: packages/cmd/src/CommandProcessor.ts
Comment (dkelosky) [reply packages/cmd/src/CommandProcessor.ts:821]: This one is tricky.  I spent some time with this to get it to work, and ideally, I'd like this censor logic to be in one place.  

`LoggerUtils.isSpecialValue` has a copied list of fields from [here](https://github.com/zowe/imperative/blob/master/packages/rest/src/session/ConnectionPropsForSessCfg.ts#L294).  This copied list is what we fall back to for v1 so we "support"  `--show-inputs-only` for both versions.  

I'm not clear if `LoggerUtils.isSpecialValue` is a generic API.  It adds a `.` to the special property name:
```typescript
        for (const v of specialValues) {
            if (prop.endsWith(`.${v}`)) return true;
        }
```
I'm also unclear where but with debug printing messages, I get `protocol` in the output when I print `specialValues`:
```
[
  'user',
  'password',
  'tokenValue',
  'keyPassphrase',
  'token-value',
  'base-path',
  'protocol',
  'token'
]
```

Unless there is something obvious I'm missing, I'd like to leave reconciling the secure arrays as a future tech debt item to resolve apart from this enhancement. 

Side note, idk if I'm reading this correctly, but is this logic correct?:
```typescript
            if (sec && typeof sec !== "object" && !LoggerUtils.isSpecialValue(prop))
                newData = newData.replace(new RegExp(sec, "gi"), LoggerUtils.CENSOR_RESPONSE);
```

The last condition reads like: "if not is special value then replace data"?  It seems like it would be, "if is special value then censor".

File: CHANGELOG.md
Comment (t1m0thyj) [inline CHANGELOG.md:7]: `Introduce` -> `Introduced` - please use past tense for consistency with other changelog messages 🙂

File: __tests__/__integration__/cmd/__tests__/integration/cli/gen-help/__snapshots__/Cmd.cli.gen-help.allowable-options.integration.test.ts.snap
Comment (t1m0thyj) [inline __tests__/__integration__/cmd/__tests__/integration/cli/gen-help/__snapshots__/Cmd.cli.gen-help.allowable-options.integration.test.ts.snap:76]: Would it be ideal for the "Global Options" to be listed in alphabetical order? I know they already were out of order with `--rfj` coming before `-h`, but now that we're adding more options to the list it strikes me as odd that they have no clear order.

File: CHANGELOG.md
Comment (dkelosky) [reply CHANGELOG.md]: fixe**D**

File: __tests__/__integration__/cmd/__tests__/integration/cli/gen-help/__snapshots__/Cmd.cli.gen-help.allowable-options.integration.test.ts.snap
Comment (dkelosky) [reply __tests__/__integration__/cmd/__tests__/integration/cli/gen-help/__snapshots__/Cmd.cli.gen-help.allowable-options.integration.test.ts.snap:76]: I think alphabetical menu is a good idea if universally applied.  `ZOSMF CONNECTION OPTIONS`, `RESPONSE FORMAT OPTIONS`, and `OPTIONS` are also non-alphabetical.  

`CommandPreparer.appendAutoOptions` occurs in logic after the new `CommandPreparer.appendPassOnOptions`.  Perhaps this could be adjusted, but I'm unsure of the implications.  `passOn` has a `merge` behavior and changing the ordering might be non-trivial.  Is it acceptable to address menu organization as a separated issue?

Comment (t1m0thyj) [reply __tests__/__integration__/cmd/__tests__/integration/cli/gen-help/__snapshots__/Cmd.cli.gen-help.allowable-options.integration.test.ts.snap:76]: Good points, changing the order could be non-trivial and seems outside the scope of this PR. Created #761 🙂

File: packages/cmd/src/CommandProcessor.ts
Comment (zFernand0) [reply packages/cmd/src/CommandProcessor.ts:821]: > It adds a . to the special property name

that's v2 only, for v1 I guess we could replace the . with a _ 🙂
> I'd like to leave reconciling the secure arrays as a future tech debt item to resolve apart from this enhancement

That's fine by me (though we may never get to it) 😛
> The last condition reads like: "if not is special value then replace data"? It seems like it would be, "if is special value then censor".

Definitely feels like a bug unless the isSpecialValue give backward results 😛

Comment (dkelosky) [reply packages/cmd/src/CommandProcessor.ts:821]: Thanks for reviewing @zFernand0 !

**Comment by dkelosky** [comment]
Yo - can you guys let me know when would be a good time would be a good time to get this reviewed?  I'll delay `CHANGELOG.md`  conflict resolution until then.  

It's kind of a pain to keep resolving the same conflict only to need to do it again a few days later.  Is there some process improvement here that would make contributions more seamless?  I don't have any good ideas, but I think this current approach might discourage contributions.

File: packages/cmd/src/CommandProcessor.ts
Comment (t1m0thyj) [inline packages/cmd/src/CommandProcessor.ts:820]: I think `this.mConfig.api.secure.securePropsForProfile` should be used here instead. Getting the list of all secure fields in the file may have undesired behavior. For example with the following config:
```
{
    "$schema": "./zowe.schema.json",
    "profiles": {
        "lpar1": {
            "type": "zosmf",
            "properties": {
                "host": "lpar1.broadcom.net",
                "port": 443
            },
            "secure": []
        },
        "lpar2": {
            "type": "zosmf",
            "properties": {
                "port": 1443
            },
            "secure": [
                "host"
            ]
        },
        "base": {
            "type": "base",
            "properties": {
                "rejectUnauthorized": true
            },
            "secure": [
                "user",
                "password"
            ]
        }
    },
    "defaults": {
        "zosmf": "lpar1",
        "base": "base"
    },
    "autoStore": true
}
```

Here the host property is only secure in the lpar2 profile, so running `zowe jobs ls jobs --show-inputs-only` should show the host property in plain text that is present in the default lpar1 profile. Currently I see `host: (secure value)` instead.

**Review by gejohnston — APPROVED** [review]

File: packages/cmd/src/CommandProcessor.ts
Comment (dkelosky) [reply packages/cmd/src/CommandProcessor.ts:820]: your point makes sense - i suppose the same will happen if a base profile marks `host` secure while zosmf does not; i think `host` will be in the secure list.  the undesired behavior is that we may mistakenly show something as `(secure value)` when it is not.

one struggle i have - i'm not clear how to determine the name of the profile used from `CommandProcessor`.  i think we had this problem before in command handlers: we have access to the profile contents / values and type but not the profile name to use `securePropsForProfile`.  do you see some way to determine this?

File: CHANGELOG.md
Comment (dkelosky) [reply CHANGELOG.md]: side note - and I'm being a little difficult here - if this is a standard it should be documented in `contributing.md`.  if it is, it's fine to link it.  otherwise, these things feel arbitrary and unknowable to casual contributors.

File: packages/cmd/src/CommandProcessor.ts
Comment (t1m0thyj) [reply packages/cmd/src/CommandProcessor.ts:820]: Sorry I missed this comment until now, I believe `ConfigUtils.getActiveProfileName` might help here: https://github.com/zowe/imperative/blob/next/packages/config/src/ConfigUtils.ts#L49

Comment (dkelosky) [reply packages/cmd/src/CommandProcessor.ts]: side note, i noticed, `ConfigUtils.getActiveProfileName` returns `host` as a secure prop for `zosmf2` even though it is not:
```
        "lpar1": {
            "properties": {
                "host": "usilca11.lvn.broadcom.net"
            },
            "profiles": {
                "zosmf": {
                    "type": "zosmf",
                    "properties": {
                        "port": 1443
                    },
                    "secure": [
                        "host"
                    ]
                },
                "zosmf2": {
                    "type": "zosmf",
                    "properties": {
                        "port": 1443
                    },
                    "secure": []
                },
```

in any case, this is only a display command option to help users.  the only impact in worst case scenario is that we show a field as `(secure value)` when it is not.  it also seems unlikely in both of our examples that a user would want a field like `host` secure for one `zosmf` profile and not another.  

no action needed - i'm just recording considerations for this discussion

Comment (gejohnston) [reply packages/cmd/src/CommandProcessor.ts]: I thought that a PR was recently merged which resolved an issue about returning host as secure when it is only secure for some profiles. Perhaps after you merge next into this branch, the problem will go away?

**Review by t1m0thyj — APPROVED** [review]

File: packages/cmd/src/CommandProcessor.ts
Comment (t1m0thyj) [reply packages/cmd/src/CommandProcessor.ts]: I believe the latest changes from next have been merged in, if this behavior persists I think we should file it as an issue. 🙂

Comment (dkelosky) [reply packages/cmd/src/CommandProcessor.ts]: correction, it's `this.mConfig.api.secure.securePropsForProfile` instead of `ConfigUtils.getActiveProfileName` that causes an issue.  

it still happens to me after syncing with `next` this morning.  i dont get an issue using @t1m0thyj's sample profile with `lpar1.zosmf` and `lpar2.zosmf`, but i do get an issue with `lpar1.zosmf` and `lpar1.zosmf2` as in my example.  

two instances of zosmf on the sample lpar2 don't make sense, but i the `lpar` higher level is an arbitrary construction.  

Here is the code with trace messages added:
```typescript
    public securePropsForProfile(profileName: string) {
        console.log(`>>>> entry <<<<`)
        const profilePath = this.mConfig.api.profiles.expandPath(profileName);
        console.log(`profilePath: ${profilePath}`)
        const secureProps = [];
        for (const propPath of this.secureFields()) {
            console.log(`propPath: ${propPath}`)
            const pathSegments = propPath.split(".");  // profiles.XXX.properties.YYY
            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
            if (profilePath.startsWith(pathSegments.slice(0, -2).join("."))) {
                secureProps.push(pathSegments.pop());
            }
        }
        console.log(`secureProps ${secureProps}`)
        console.log(`>>>> exit <<<<`)
        return secureProps;
    }
```

here's the command i run and output I see:
```
C:\dev\systems\ca11>zowe jobs list jobs --show-inputs-only --zosmf-p lpar1.zosmf2
>>>> entry <<<<
profilePath: profiles.lpar1.profiles.zosmf2
propPath: profiles.lpar1.profiles.zosmf.properties.host
propPath: profiles.my_base.properties.user
propPath: profiles.my_base.properties.password
secureProps host
>>>> exit <<<<
>>>> entry <<<<
profilePath: profiles.my_base
propPath: profiles.lpar1.profiles.zosmf.properties.host
propPath: profiles.my_base.properties.user
propPath: profiles.my_base.properties.password
secureProps user,password
>>>> exit <<<<
```

it looks like `zosmf2` becomes `zosmf` in `ConfigProfiles.expandPath`

**Pull request merged**
